gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java

```diff
@@ -4049,8 +4061,25 @@ public boolean isSnapshot() {
   }
 
   @Override
-  public void recordVersionForSnapshot(Object member, long version, Region region ) {
+  public void recordVersionForSnapshot(Object member, long version, Region region) {
     queue.add(new VersionInformation(member, version, region));
+    Boolean wasPresent = writeRegions.putIfAbsent(region, true);
+    // Handle both row buffer and column tables together , when column  table
+    // insert is transactional , both of them can have data inserted from a transaction
+    // and it might lead to deadlock.
+    if (wasPresent == null) {
+      if (region instanceof BucketRegion) {
+        BucketRegion br = (BucketRegion)region;
+        BucketRegion companionRegion = br.companionRegion();
+        if (companionRegion != null) {
```

Comment:
Hmm. I am changing it to take lock on row buffer. cached batch table will also use row buffer locks both read and write.